osutil/sync_dir_test.go

```diff
@@ -0,0 +1,175 @@
+}
+
+func (s *EnsureDirStateSuite) TestIgnoresUnrelatedFiles(c *C) {
+	name := filepath.Join(s.dir, "unrelated")
+	err := ioutil.WriteFile(name, []byte(`text`), 0600)
+	c.Assert(err, IsNil)
+	changed, removed, err := osutil.EnsureDirState(s.dir, s.glob, map[string]*osutil.FileState{})
+	c.Assert(err, IsNil)
+	// Report says that nothing has changed
+	c.Assert(changed, HasLen, 0)
+	c.Assert(removed, HasLen, 0)
+	// The file is still there
+	_, err = os.Stat(name)
+	c.Assert(err, IsNil)
+}
+
+func (s *EnsureDirStateSuite) TestCorrectsFilesWithDifferentSize(c *C) {
+	name := filepath.Join(s.dir, "differing.snap")
+	err := ioutil.WriteFile(name, []byte(``), 0600)
+	c.Assert(err, IsNil)
+	changed, removed, err := osutil.EnsureDirState(s.dir, s.glob, map[string]*osutil.FileState{
+		"differing.snap": {Content: []byte(`Hello World`), Mode: 0600},
+	})
+	c.Assert(err, IsNil)
+	// changed file is reported
+	c.Assert(changed, DeepEquals, []string{"differing.snap"})
+	c.Assert(removed, HasLen, 0)
+	// The content is changed
+	content, err := ioutil.ReadFile(name)
+	c.Assert(err, IsNil)
+	c.Assert(content, DeepEquals, []byte(`Hello World`))
+	// The permissions are what we expect
+	stat, err := os.Stat(name)
+	c.Assert(err, IsNil)
+	c.Assert(stat.Mode().Perm(), Equals, os.FileMode(0600))
+}
+
+func (s *EnsureDirStateSuite) TestCorrectsFilesWithSameSize(c *C) {
+	name := filepath.Join(s.dir, "differing.snap")
+	err := ioutil.WriteFile(name, []byte(`evil`), 0600)
+	c.Assert(err, IsNil)
+	changed, removed, err := osutil.EnsureDirState(s.dir, s.glob, map[string]*osutil.FileState{
+		"differing.snap": {Content: []byte(`good`), Mode: 0600},
+	})
+	c.Assert(err, IsNil)
+	// changed file is reported
+	c.Assert(changed, DeepEquals, []string{"differing.snap"})
+	c.Assert(removed, HasLen, 0)
+	// The content is changed
+	content, err := ioutil.ReadFile(name)
+	c.Assert(err, IsNil)
+	c.Assert(content, DeepEquals, []byte(`good`))
+	// The permissions are what we expect
+	stat, err := os.Stat(name)
+	c.Assert(err, IsNil)
+	c.Assert(stat.Mode().Perm(), Equals, os.FileMode(0600))
+}
+
+func (s *EnsureDirStateSuite) TestFixesFilesWithBadPermissions(c *C) {
+	name := filepath.Join(s.dir, "sensitive.snap")
+	// NOTE: the file is wide-open for everyone
+	err := ioutil.WriteFile(name, []byte(`password`), 0666)
+	c.Assert(err, IsNil)
+	changed, removed, err := osutil.EnsureDirState(s.dir, s.glob, map[string]*osutil.FileState{
+		// NOTE: we want the file to be private
+		"sensitive.snap": {Content: []byte(`password`), Mode: 0600},
+	})
+	c.Assert(err, IsNil)
+	// changed file is reported
+	c.Assert(changed, DeepEquals, []string{"sensitive.snap"})
+	c.Assert(removed, HasLen, 0)
+	// The content is still the same
+	content, err := ioutil.ReadFile(name)
+	c.Assert(err, IsNil)
+	c.Assert(content, DeepEquals, []byte(`password`))
+	// The permissions are changed
+	stat, err := os.Stat(name)
+	c.Assert(err, IsNil)
+	c.Assert(stat.Mode().Perm(), Equals, os.FileMode(0600))
+}
```

Comment:
Since EnsureDirState is a loop, I think it would be wise to add a few tests for when the map is > 1. Eg: 2 files with no error, 1st file with error and 2nd ok, 1st file ok and 2nd error, etc.